bundles/org.openhab.binding.flicbutton/src/main/java/org/openhab/binding/flicbutton/handler/FlicButtonHandler.java

```diff
@@ -0,0 +1,177 @@
+
+import java.io.IOException;
+import java.util.Objects;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.openhab.core.library.types.DecimalType;
+import org.openhab.core.thing.*;
+import org.openhab.core.types.Command;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.flic.fliclib.javaclient.BatteryStatusListener;
+import io.flic.fliclib.javaclient.Bdaddr;
+import io.flic.fliclib.javaclient.ButtonConnectionChannel;
+import io.flic.fliclib.javaclient.enums.ConnectionStatus;
+import io.flic.fliclib.javaclient.enums.DisconnectReason;
+
+/**
+ * The {@link FlicButtonHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Patrick Fink - Initial contribution
+ */
+public class FlicButtonHandler extends ChildThingHandler<FlicDaemonBridgeHandler> {
+
+    private Logger logger = LoggerFactory.getLogger(FlicButtonHandler.class);
+    private ScheduledFuture<?> delayedDisconnectTask;
+    private DisconnectReason latestDisconnectReason;
+    private ButtonConnectionChannel eventConnection;
+    private Bdaddr bdaddr;
+    BatteryStatusListener batteryConnection;
+
+    public FlicButtonHandler(Thing thing) {
+        super(thing);
+    }
+
+    public Bdaddr getBdaddr() {
+        return bdaddr;
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        // Pure sensor -> no commands have to be handled
+    }
+
+    @Override
+    public void initialize() {
+        super.initialize();
+        bdaddr = new Bdaddr((String) this.getThing().getConfiguration().get(CONFIG_ADDRESS));
+        if (bridgeValid) {
+            initializeThing();
+        }
+    }
+
+    public void initializeThing() {
+        try {
+            FlicButtonBatteryLevelListener batteryListener = new FlicButtonBatteryLevelListener(this);
+            BatteryStatusListener batteryConnection = new BatteryStatusListener(getBdaddr(), batteryListener);
+            bridgeHandler.getFlicClient().addBatteryStatusListener(batteryConnection);
+            this.batteryConnection = batteryConnection;
+
+            FlicButtonEventListener eventListener = new FlicButtonEventListener(this);
+            synchronized (eventListener) {
+                ButtonConnectionChannel eventConnection = new ButtonConnectionChannel(getBdaddr(), eventListener);
+                bridgeHandler.getFlicClient().addConnectionChannel(eventConnection);
+                this.eventConnection = eventConnection;
+                eventListener.wait(5000);
+                // Listener calls initializeStatus() before notifying so that ThingStatus is set at this point
+            }
+        } catch (IOException | InterruptedException e) {
+            logger.info("Connection setup for Flic Button {} failed.", this.getThing(), e);
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "Connection setup failed");
```

Comment:
The log message could be removed, as the status update is already logged by the framework.

You might want to include the exception's message into the Thing status detail message.